quarkchain/cluster/tests/test_failed_transactions_gas_fee.py

```diff
@@ -0,0 +1,112 @@
+
+    def test_failed_transaction_gas(self):
+        """in-shard revert contract transaction validating the failed transaction gas used
+        """
+        QKC = token_id_encode("QKC")
+        id1 = Identity.create_random_identity()
+        acc1 = Address.create_from_identity(id1, full_shard_key=0)
+        acc2 = Address.create_random_account(full_shard_key=0)
+
+        env = get_test_env(
+            genesis_account=acc1,
+            genesis_minor_token_balances={
+                self.GENESIS_TOKEN: 200 * 10 ** 18,
+            },
+        )
+
+        state = create_default_shard_state(env=env)
+
+        # 1. creat failed contract with revert operation
+
+        CONTRACT_CREATION_WITH_REVERT_BYTECODE = "6080604052348015600f57600080fd5b503373ffffffffffffffffffffffffffffffffffffffff167fa9378d5bd800fae4d5b8d4c6712b2b64e8ecc86fdc831cb51944000fc7c8ecfa60405160405180910390a2600080fdfe"
+        """
+        pragma solidity ^0.5.1;
+        contract RevertContract {
+            event Hi(address indexed);
+            constructor() public {
+                emit Hi(msg.sender);
+                revert();
+            }
+            function f() public {
+                emit Hi(msg.sender);
+            }
+        }
+        """
+        # This transaction cost is calculated by remix, which is smaller than the opcodes.GTXCOST due to revert.
+        FAILED_TRANSACTION_COST = 60004
+    
+        tx = contract_creation_tx(
+            shard_state=state,
+            key=id1.get_key(),
+            from_address=acc1,
+            to_full_shard_key=acc1.full_shard_key,
+            bytecode=CONTRACT_CREATION_WITH_REVERT_BYTECODE,
+            gas_token_id=self.genesis_token,
+            transfer_token_id=self.genesis_token,
+        )
+        # Should succeed
+        self.assertTrue(state.add_tx(tx))
+        b1 = state.create_block_to_mine(address=acc2)
+        self.assertEqual(len(b1.tx_list), 1)
+
+        state.finalize_and_add_block(b1)
+
+        self.assertEqual(state.header_tip, b1.header)
+
+        self.assertNotEqual(FAILED_TRANSACTION_COST , opcodes.GTXCOST)
+
+        # Make sure the FAILED_TRANSACTION_COST is consumed by the sender
+        self.assertEqual(
+            state.get_token_balance(id1.recipient, self.genesis_token),
+            200 * 10 ** 18 - FAILED_TRANSACTION_COST,
+        )
+
+        self.assertNotEqual(
+            state.get_token_balance(id1.recipient, self.genesis_token),
+            200 * 10 ** 18 - opcodes.GTXCOST,
+        )
+
+        # Make sure the accurate gas fee is obtained by the miner
+        self.assertEqual(
+            state.get_token_balance(acc2.recipient, self.genesis_token),
+            self.getAfterTaxReward(FAILED_TRANSACTION_COST + self.shard_coinbase),
+        )
+
+        self.assertNotEqual(
+            state.get_token_balance(acc2.recipient, self.genesis_token),
+            self.getAfterTaxReward(opcodes.GTXCOST + self.shard_coinbase),
```

Comment:
can you help changing the naming to lower case? I know it's also named in `test_shard_state.py`.  `getAfterTaxReward` -> `get_after_tax_reward`